### PR TITLE
feat: support deployment volumes and container volumeMounts.

### LINKED ISSUE #1417: Support Volume and VolumeMounts in EnvoyProxy K8s Deployment spec
*Description*:
Add the ability to configure `volumes` and `volumeMounts` in the managed Envoy Proxy data plane using the [EnvoyProxy API](https://github.com/envoyproxy/gateway/blob/ad58babaf83cf9643345ad10bd5be0f22c147258/api/config/v1alpha1/envoyproxy_types.go#L77)
https://kubernetes.io/docs/concepts/storage/volumes/

Use Case
* Add WASM modules into Envoy Proxy

### PR DESCRIPTION
**Pull request by qicz**
Fixes #1417

### CODE REVIEW

**Review by Xunzhuo — COMMENTED** [review]
Great work, can you add a doc for it along with this PR?

**Review by Xunzhuo — COMMENTED** [review]
So this PR adds volumes support on both eg deployment and ratelimit deployment ? @qicz

**Comment by qicz** [comment]
> So this PR adds volumes support on both eg deployment and ratelimit deployment ? @qicz

yes

**Review by Xunzhuo — DISMISSED** [review]
/lgtm

File: api/config/v1alpha1/shared_types.go [hunk 1]
```diff
@@ -83,6 +83,12 @@ type KubernetesPodSpec struct {
 	// If specified, the pod's tolerations.
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// Volumes List of volumes that can be mounted by containers belonging to the pod.
```
Comment (arkodg) [inline api/config/v1alpha1/shared_types.go:87]:
```suggestion
	// Volumes that can be mounted by containers belonging to the pod.
```

File: api/config/v1alpha1/shared_types.go [hunk 2]
```diff
@@ -109,6 +115,12 @@ type KubernetesContainerSpec struct {
 	//
 	// +optional
 	Image *string `json:"image,omitempty"`
+
+	// VolumeMounts Pod volumes to mount into the container's filesystem.
```
Comment (arkodg) [inline api/config/v1alpha1/shared_types.go:119]:
```suggestion
	// VolumeMounts are volumes to mount into the container's filesystem.
```

File: docs/latest/user/customize-envoyproxy.md [hunk 3]
```diff
@@ -160,6 +160,37 @@ EOF
...
+        container:
+          volumeMounts:
+          - mountPath: /certs
+            name: certs
+            readOnly: true
+        pod:
+          volumes:
+          - name: certs
+            secret:
+            secretName: envoy-cert
```
Comment (arkodg) [inline docs/latest/user/customize-envoyproxy.md:188]:
```suggestion
              secretName: envoy-cert   
```

File: docs/latest/user/rate-limit.md [hunk 4]
```diff
@@ -603,6 +603,9 @@ data:
             securityContext:
               runAsUser: 2000
               allowPrivilegeEscalation: false
+            volumeMounts:
```
Comment (arkodg) [inline docs/latest/user/rate-limit.md:606]: lets rm this from the docs, since we are focusing on Xds for RL config instead of File

**Comment by qicz** [comment]
@arkodg all comments resolved, ptal

File: api/config/v1alpha1/shared_types.go [hunk 5]
```diff
@@ -83,6 +83,12 @@ type KubernetesPodSpec struct {
 	// If specified, the pod's tolerations.
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// Volumes that can be mounted by containers belonging to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes
+	//
+	// +optional
+	Volumes []corev1.Volume `json:"volumes,omitempty"`
```
Comment (arkodg) [inline api/config/v1alpha1/shared_types.go:91]: should this be at the top level instead of under `pod` ?

Comment (qicz) [reply api/config/v1alpha1/shared_types.go:91]: I think should keep same level like `Affinity` at `Deployment` yaml

**Review by Xunzhuo — APPROVED** [review]

**Pull request merged**